vizro-core/tests/unit/vizro/models/test_utils.py

```diff
@@ -0,0 +1,248 @@
+import textwrap
+from typing import List, Optional
+
+import pytest
+import vizro.models as vm
+import vizro.plotly.express as px
+from vizro.actions import export_data
+from vizro.models._utils import _concatenate_code
+from vizro.models.types import capture
+
+extra_imports_string = "import vizro.models as vm\nimport pandas as pd"
+code_string = "vm.Card(text='Foo')"
+callable_defs_string = """def f():\n  return 'hi'"""
+data_settings_string = """# data_manager["iris"] = ===> Fill in here <==="""
+
+expected_assembled_linted_code = textwrap.dedent(
+    """############ Imports ##############
+import vizro.models as vm
+
+
+####### Function definitions ######
+def f():
+    return "hi"
+
+
+####### Data Manager Settings #####
+#######!!! UNCOMMENT BELOW !!!#####
+# from vizro.managers import data_manager
+# data_manager["iris"] = ===> Fill in here <===
+
+
+########### Model code ############
+vm.Card(text="Foo")
+"""
+)
+
+
+class TestCodeConcatenation:
+    def test_concatenate_code(self):
+        result = _concatenate_code(
+            code=code_string,
+            extra_imports=extra_imports_string,
+            callable_defs=callable_defs_string,
+            data_settings=data_settings_string,
+        )
+        assert result == expected_assembled_linted_code
+
+
+@capture("graph")
+def chart(data_frame, hover_data: Optional[List[str]] = None):
+    return px.bar(data_frame, x="sepal_width", y="sepal_length", hover_data=hover_data)
+
+
+@pytest.fixture
+def page_pre_defined_actions():
+    return vm.Page(
+        title="Page 1",
+        components=[
+            vm.Graph(figure=px.bar("iris", x="sepal_width", y="sepal_length")),
+            vm.Button(
+                text="Export data",
+                actions=[
+                    vm.Action(function=export_data()),
+                    vm.Action(function=export_data()),
+                ],
+            ),
+        ],
+    )
+
+
+@pytest.fixture
+def chart_dynamic():
+    function_string = textwrap.dedent(
+        """
+        @capture("graph")
+        def chart_dynamic(data_frame, hover_data: Optional[List[str]] = None):
+            return px.bar(data_frame, x="sepal_width", y="sepal_length", hover_data=hover_data)
+        """
+    )
+
+    local_scope = {}
+    exec(function_string, globals(), local_scope)
+    chart_dynamic = local_scope["chart_dynamic"]
+    return chart_dynamic
```

Comment:
```suggestion
    exec(function_string)
    return locals()["chart_dynamic"]
```

Not 100% sure but I think this will do what you're looking for a bit more simply and conventionally.

Reply:
Actually I have no idea why we're doing `exec` here at all? 🤔 What am I missing? Why can't we just return the `chart_dynamic` function? We don't need it in a fixture if that's making it more awkward?

Reply:
We want to test the case where the source code does not exist. At first I thought of mocking it, but I think it is even better to test a realistic case. Wdyt?

Reply:
Applied your changes though, thanks!

Reply:
Ohhh I see, so the whole point here is that if you write this out as a "normal" function then it will exist in the source code of test_utils.py and hence you won't be able to test the "code is not available" case. This makes sense now.

The comment in the test is good but if you added something to this fixture just to explain what it's doing then that would be great.